pkg/upgraders/postupgradestep.go

```diff
@@ -0,0 +1,111 @@
+package upgraders
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/openshift/managed-upgrade-operator/config"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	fioNamespace    string = "openshift-file-integrity"
+	fioObject       string = "osd-fileintegrity"
+	frOCMBaseDomain string = "openshiftusgov.com"
+)
+
+var reinitAnnotation = map[string]string{"file-integrity.openshift.io/re-init": ""}
+
+// ugConfigManager stores the configManager section of the upgradeconfig manager configmap
+type ugConfigManager struct {
+	Source     string `yaml:"source"`
+	OcmBaseURL string `yaml:"ocmBaseUrl"`
+}
+
+// ugConfigManagerSpec stores the config.yaml section of the upgradeconfig manager configmap
+type ugConfigManagerSpec struct {
+	Config ugConfigManager `yaml:"configManager"`
+}
+
+// PostUpgradeProcedures are any misc tasks that are needed to be completed after an upgrade has finished to ensure healthy state
+// Currently the only task is to reinit file integrity operator due to changes that come from upgrades
+func (c *clusterUpgrader) PostUpgradeProcedures(ctx context.Context, logger logr.Logger) (bool, error) {
+
+	frCluster, err := c.frClusterCheck(ctx)
+	if err != nil {
+		return false, err
+	}
+	if !frCluster {
+		logger.Info("Non-FedRAMP environment...skipping PostUpgradeFIOReInit ")
+		return true, nil
+	}
+	err = c.postUpgradeFIOReInit(ctx, logger)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// frClusterCheck checks to see if the upgrading cluster is a FedRAMP cluster to determine if we need to re-init the File Integrity Operator
+func (c *clusterUpgrader) frClusterCheck(ctx context.Context) (bool, error) {
+	ocmConfig := &corev1.ConfigMap{}
+	err := c.client.Get(context.TODO(), client.ObjectKey{Namespace: config.OperatorNamespace, Name: config.ConfigMapName}, ocmConfig)
```

Comment:
not my space, so feel free to ignore me if I'm not making sense.  But wouldn't it make sense to poke the FIO post-update whenever it was installed, on both FedRAMP and otherwise?

Reply:
Hey @wking I'm sorry, I'm not sure I follow. Do you mean why not poke FIO no matter what kind of cluster it is, whether FedRAMP or not? This is currently only set for FedRAMP as FIO is not deployed in commercial OSD/ROSA clusters by default where in FedRAMP it is. 

 In FedRAMP its part of the ROSA deployment and used by our team as a requirement of FedRAMP for us to monitor. If a customer decides to use this tool in commercial, while it might be beneficial to add that functionality, they may not all desire us messing with their security configurations? I could see the argument for either way, but this update is targeting specifically only the SREP managed FIO deployments, since customers can deploy to any namespace and name the actual `FileIntegrity` object anything they want. It could be done but not sure if we want that default behavior? Thanks for the input!

Reply:
>  I could see the argument for either way, but this update is targeting specifically only the SREP managed FIO deployments...

It feels like there should be some way to look at a FileIntegrity resource and decide it its yours or not.  Perhaps by setting [`ownerReferences`][1] for SRE-P-owned resources, and then switching on the presence of those references.  Then "is this my FileIntegrity, or did someone else create a FileIntegrity resource called `osd-fileintegrity` in the `openshift-file-integrity` namespace?" can be answered by inspecting the resource itself.  And really, those naming conventions already seem so SRE-P specific that I would be very surprised to find someone else's FileIntegrity resource by mistake.  But sure, if we think that checking a ConfigMap to decide if we're FedRAMP is more reliable than using something about the FileIntegrity resource itself, then go for it :)

[1]: https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications

Reply:
I have to admit this PR gives me a few not-comfortable feelings from the perspective of how this brings in a (indirect, as API types aren't being imported) dependency between the two operators, and maintenance of things like a Fedramp domain name in MUO code.

I'd wish that the FIO had a controller that watched for UpgradeConfigs (perhaps added a finalizer to them, too) and did its thing based on that. If the UpgradeConfig is being deleted and its Status indicates that it completed an upgrade -> reinit whatever needs initializing and remove the finalizer. That divorces MUO's need to have any knowledge.

But given that FIO is an OCP operator and not an OSD/ROSA-specific operator I don't know how likely that change is.

Reply:
I tried pulling in the bits needed from the FIO code base to check for the file integrity API's needed but go versions and dependency differences cause a lot of issues and wouldn't work, so I ended up using the unstructured library. Its been months since I last tried but they are on go 1.16 so not sure it'll work still. I can give it another test if desired.

The issue with adding any MUO specific stuff to FIO is that not just openshift customers may be leveraging FIO since its in OperatorHub. Its a pretty universal tool that could work with any K8s really, so adding something specific to OpenShift might not go well with other users.

Reply:
are the triggers cluster-scoped or node-scoped?  Because ideally, you'd trigger a node scan after each node change.  [RFE-1925][1] is tracking node-update hooks, which would give you one place to plug in.  On the other hand, it's an RFE because it hasn't been implemented yet.  Tricky problem.

[1]: https://issues.redhat.com/browse/RFE-1925

Reply:
When the annotation is applied to the `FileIntegrity` it triggers the daemonset pods to force reinit their database so a single annotation triggers all nodes to update their DB files so that the current file contents are the expected baseline.